Better api usage and better retry

sebageek opened the pull request:
Lots of api usage improvements, better retry logic and token renewal.

sebageek committed:
Improve performance by direct queries for BD/EPG

To get an EPG or BD the driver now constructs the DN and directly
fetches it from the device. For EPG, the subtree classes are also
limited to what is actually needed. In comparison to querying the full
tenant each time this is a performance increase from ~600-900ms to
~30-60ms.

This change also makes the driver's use of the acicobra v3 compatible
with ACI 4.2, as getting a Tenant with subtree=full contains API objects
that are not understood by acicobra 3 and we no longer query these
objects.

sebageek committed:
Reimplement and make consistent use of retry logic

A retry needs to be done when the APIC does not want to talk to us
anymore. This might be caused on an internal error, a locked config
object, a broken APIC (and we need to switch over) or due to an auth
token timeout.

Before this commit retrying was only used inconsistently for certain
queries, now it is used for all queries done by the driver.

sebageek committed:
Improve get_all_epgs() performance with better filtering

get_all_epgs() previously iterated over all tenants, then got all EPGs
of that tenant. As it got everything without subtree=full, this took
around 8s in qa.

Now we just get all EPGs from the ACI that have a tenant starting with
the preconfigured tenant-prefix, which reduces the query time in this
instance to 200ms.

sebageek committed:
Renew APIC token before it times out

We don't need to wait until our token times out and a query failure
triggers a relogin. We can just look at our token timeout and see if it
is still valid. If it is bound to timeout in the next 120s
(configurable) then we call reauth() to refresh it. If it's invalid we
trigger a relogin, but this should only happen when the syncloop is
turned off or something else hangs.

grandchild commented:
Commit message in _Improve get_all_epgs() performance with better filtering_:

> [...] this could be done in ~8s in qa.

This sounds like a good thing, but it's not. How about

> [...] this took around 8s in QA.

?

sebageek committed:
Return None on not found for single-mode in cobra client

When called with single=True, query() unpacks the result. If the result
is empty then we will return None instead of an empty list.

For getting a BD/EPG we want to have at least a debug log message so we
throw a runtime error that we then catch and log.
